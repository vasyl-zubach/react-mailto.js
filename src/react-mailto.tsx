import React, { HTMLAttributes, SyntheticEvent } from 'react';
import window from 'window-or-global';

import { Props } from './types';

const encode = encodeURIComponent;

const getMailtoLink = function getMailtoLink({
  subject,
  to,
  body,
  cc,
  bcc
}: Partial<Props>) {
  const link: (string | null)[] = [
    cc ? `cc=${cc}` : null,
    bcc ? `bcc=${bcc}` : null,
    subject ? `subject=${encode(subject)}` : null,
    body ? `body=${encode(body)}` : null
  ].filter(Boolean);

  return `${to}?${link.join('&')}&_c=${Date.now()}`;
};

const handleSecureClick =
  (to: string, onClick?: (event: SyntheticEvent<HTMLAnchorElement>) => void) =>
  (event: SyntheticEvent<HTMLAnchorElement>) => {
    event.preventDefault();
    onClick?.(event);

    window?.location?.assign?.(to);
  };

export const Mailto = ({
  to,
  cc,
  bcc,
  subject,
  body,
  secure,
  ...props
}: Props & HTMLAttributes<HTMLAnchorElement>) => {
  const link = getMailtoLink({ to, cc, bcc, subject, body });
  const isSecure = secure === true;
  const href = `mailto:${link}`;
  const onClick = isSecure
    ? handleSecureClick(href, props.onClick)
    : props.onClick;
  return <a {...props} href={isSecure ? '' : href} onClick={onClick} />;
};
