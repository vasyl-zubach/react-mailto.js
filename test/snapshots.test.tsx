import { expect, test } from '@jest/globals';

import React from 'react';

import renderer from 'react-test-renderer';

import * as Stories from '../stories/react-mailto.story';

global.Date.now = () => 1234567890;

Object.keys(Stories)
  .filter((i) => i !== 'default')
  .forEach((storyTitle) => {
    test(`Story: ${storyTitle}`, () => {
      const Component = Stories[storyTitle];
      const tree = renderer.create(<Component />);
      expect(tree).toMatchSnapshot();
    });
  });
