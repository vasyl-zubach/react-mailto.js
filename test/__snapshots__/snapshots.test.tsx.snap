// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Story: additionalAttributes 1`] = `
<a
  data-additional="test"
  href="mailto:vasyl@example.com?&_c=1234567890"
>
  Contact us
</a>
`;

exports[`Story: basicConfig 1`] = `
<a
  href="mailto:vasyl@example.com?subject=Some%20email%20subject&body=This%20is%20the%20body%20of%20your%20message%0AMultilined%20also&_c=1234567890"
>
  Email us
</a>
`;

exports[`Story: customStyles 1`] = `
<a
  href="mailto:vasyl@example.com?&_c=1234567890"
  style={
    {
      "border": "1px #000 solid",
    }
  }
>
  Contact us
</a>
`;

exports[`Story: maximumConfig 1`] = `
<a
  href=""
  onClick={[Function]}
>
  Email us
</a>
`;
